### DISCUSSION TITLE
Outbound TCP port limitations on GitHub actions runners?

### DISCUSSION THREAD

**Discussion by MichaIng** [community/community#31144] [Actions] [answered]
Hey guys,

recently we switched our SSH server from port 22 to a random 5 digits port to reduce random connections/attacks, spamming our logs and keeping fail2ban busy.

There are GitHub workflows which do uploads to this SSH server. I updated them to upload to the new port, but get a
```
curl: (60) SSL peer certificate or SSH remote key was not OK
```
as response, using `curl` for SFTP uploads.

From any other machine outside of GitHub runners, the same uploads work perfectly fine, and switching back to port 22 fixes uploads for the GitHub workflows as well, so I assume that such high port numbers are somehow blocked or redirected for outgoing requests, making it impossible to use them.

Can anyone verify this, is it documented somewhere (I couldn't find anything about it), basically which ports can I use for SFTP uploads from a GitHub runner?

Best regards in advance!

**Comment by airtower-luna** [comment, marked as answer]
How did you provide the correct public key to curl? With OpenSSH in general a key is accepted (or not) for a host/port combination, so you might need to adjust that, too.

**Reply by MichaIng** [reply]
Via `~/.ssh/known_hosts`, containing the hostname only, not the port. On all other systems, uploads to the server work well with `known_hosts` containing hostname only, but I'll now try to define it explicitly in the GitHub workflow.

**Reply by MichaIng** [reply]
This does not work, when I add `my.host.name:port key-type key` to `known_hosts`, uploads generally fail on all systems, so I suspect it is not intended to contain a port. Removing the port fixes uploads on a local system here with `curl -T filename sftp://user@my.host.name:port/path` also with any port different than 22.

Btw, GitHub runners are Ubuntu Jammy and my other local systems are Debian Buster to Bookworm, but there shouldn't be a difference, is it? Just checked, `curl` on Debian Bookworm uses `libssh2-1`, on Ubuntu Jammy it uses `libssh-4`, are they probably expecting a different `known_hosts` syntax, Debian's curl/SSH lib without port and Ubuntu's curl/SSH lib with port?

**Reply by airtower-luna** [reply]
> This does not work, when I add my.host.name:port key-type key to known_hosts, uploads generally fail on all systems

That format is incorrect, see the format description in the [`sshd` manual page](https://manpages.debian.org/bullseye/openssh-server/sshd.8.en.html#SSH_KNOWN_HOSTS_FILE_FORMAT):

> `A hostname or address may optionally be enclosed within ‘[’ and ‘]’ brackets then followed by ‘:’ and a non-standard port number.`

So you'd need something like `[ssh.example.com]:12345`, that's also the format I see in my own `known_hosts` file. Also, try testing the connection with `ssh -v` to get some debug info.

> Btw, GitHub runners are Ubuntu Jammy and my other local systems are Debian Buster to Bookworm, but there shouldn't be a difference, is it?

Nah, it's OpenSSH either way. Unless you're using some exotic feature that was recently added or removed it shouldn't make a difference. A difference in allowed key algorithms would also explain the failure, but that shouldn't change if the port was the *only* thing you changed. :slightly_smiling_face:

**Reply by MichaIng** [reply]
See my edit above, it's `libssh2-1` vs `libssh-4` which indeed makes the difference: The first does not require the port in `known_hosts`, the second does.

> That format is incorrect, see the format description in the [sshd manual page](https://manpages.debian.org/bullseye/openssh-server/sshd.8.en.html#SSH_KNOWN_HOSTS_FILE_FORMAT):

Thanks, that was it! That format already makes sense for IPv6 support 🙂.

**Reply by airtower-luna** [reply]
> See my edit above, it's libssh2-1 vs libssh-4 which indeed makes the difference: The first does not require the port in known_hosts, the second does.

Interesting, looks like [Ubuntu has indeed changed the SSH library](https://launchpad.net/ubuntu/+source/curl/7.80.0-3) while so far [Debian hasn't](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897950). My experience is mostly with OpenSSH directly, not through curl, so I've had that port format in my `known_hosts` for years. :smile_cat: 

Glad to hear it works now, and marking the answer would be appreciated. :slightly_smiling_face:

**Reply by MichaIng** [reply]
Also great to know about this `known_hosts` format now. Since I did interactive SSH sessions on such custom port from Windows clients so far, and `libssh2-1` is not so strict about the port (without port defined, all ports are accepted, but with port defined, of course it needs to match), I didn't know about it at all 😄.

I recognised this difference of the underlying SSH library a while ago when missing ECDSA and ed25519 host key support on Debian (with curl as client), while it worked on Ubuntu.

**Reply by airtower-luna** [reply]
> I recognised this difference of the underlying SSH library a while ago when missing ECDSA and ed25519 host key support on Debian (with curl as client), while it worked on Ubuntu.

You might want to point that out in the Debian bug, that and the port thing sound like very good reasons for Debian to switch, too. :sweat_smile:

**Reply by MichaIng** [reply]
Just added it as argument here: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897950#42

Dammit, mobile email client broke plain text format 😅.
